src/main/g8/src/main/scala/worker/Worker.scala

```diff
@@ -2,95 +2,178 @@ package worker
 
 import java.util.UUID
 
-import akka.actor.SupervisorStrategy.{Restart, Stop}
-import akka.actor._
+import akka.actor.typed._
+import akka.actor.typed.scaladsl._
+import worker.Master.WorkerRequestsWork
+import worker.WorkExecutor.DoWork
+import worker.Worker.Ack
+import worker.Worker.Register
+import worker.Worker.SubmitWork
+import worker.Worker.WorkIsReady
+import worker.Worker.WorkTimeout
 
 import scala.concurrent.duration._
 
-/**
- * The worker is actually more of a middle manager, delegating the actual work
- * to the WorkExecutor, supervising it and keeping itself available to interact with the work master.
- */
-object Worker {
-
-  def props(masterProxy: ActorRef): Props = Props(new Worker(masterProxy))
-
-}
-
-class Worker(masterProxy: ActorRef)
-  extends Actor with Timers with ActorLogging {
-  import MasterWorkerProtocol._
-  import context.dispatcher
-
-
-  val workerId = UUID.randomUUID().toString
-  val registerInterval = context.system.settings.config.getDuration("distributed-workers.worker-registration-interval").getSeconds.seconds
-
-  val registerTask = context.system.scheduler.schedule(0.seconds, registerInterval, masterProxy, RegisterWorker(workerId))
-
-  val workExecutor = createWorkExecutor()
-
-  var currentWorkId: Option[String] = None
-  def workId: String = currentWorkId match {
-    case Some(workId) => workId
-    case None         => throw new IllegalStateException("Not working")
+class Worker private (workerId: String,
+                      masterProxy: ActorRef[Master.Command],
+                      ctx: ActorContext[Worker.Command],
+                      timers: TimerScheduler[Worker.Command],
+                      workExecutorFactory: () => Behavior[DoWork]) {
+  private val registerInterval = ctx.system.settings.config
+    .getDuration("distributed-workers.worker-registration-interval")
+    .toNanos
+    .nano
+
+  private val workAckTimeout = ctx.system.settings.config
+    .getDuration("distributed-workers.work-ack-timeout")
+    .toNanos
+    .nano
+
+  timers.startTimerWithFixedDelay("register", Register, registerInterval)
```

Comment:
I think we should rip out this whole registration logic as it is now covered by the receptionist. Just register all workers for a shared key and the master can subscribe to that instead. When the worker dies it is automatically deregistered from the receptionist.

Reply:
Yep will do